onnxruntime/core/framework/session_state.cc

```diff
@@ -258,45 +259,157 @@ void SessionState::CleanInitializedTensorsFromGraph() {
+          do {
+            int ort_value_idx;
+            if (st->GetOrtValueNameIdxMap().GetIdx(input_name, ort_value_idx).IsOK()) {
+              std::unordered_map<int, OrtValue>& constant_initialized_tensors = st->constant_initialized_tensors_;
+
+              if (constant_initialized_tensors.count(ort_value_idx)) {
+                bool is_packed = false;
+                const Tensor& const_initialized_tensor = constant_initialized_tensors[ort_value_idx].Get<Tensor>();
+
+                auto iter = initializers_to_share_map.find(input_name);
+                bool is_shared_initializer = (iter != initializers_to_share_map.end());
+
+                // Caching pre-packed weights is limited to shared initializers associated with the CPU EP for now
+                if (is_shared_initializer && should_cache_prepacked_weights_for_shared_initializers &&
+                    node.GetExecutionProviderType() == kCpuExecutionProvider) {  // caching of pre-packed weights' turned ON
+
+                  AllocatorPtr allocator_for_caching = prepacked_weights_container_->GetAllocator(CPU);
+                  ORT_ENFORCE(allocator_for_caching.get() != nullptr);
+
+                  PrePackedWeights weights_to_be_filled_in;
+                  // The reason we invoke PrePack() before looking into the container for any pre-packed weight
+                  // cached by another instance of the same op_type (for the ame constant initializer) is because
+                  // to truly know if we can use a cached pre-packed weight, we would hav to compare the cached pre-packed
+                  // weight with the pre-packed weight generated by this instance of the same op_type because other static
+                  // properties of the node like node attributes could play a role in the pre-packed weights' contents.
+                  ORT_RETURN_IF_ERROR(kernel->PrePack(const_initialized_tensor, input_idx, is_packed,
+                                                      &weights_to_be_filled_in,
+                                                      allocator_for_caching));
+
+                  if (is_packed) {
+                    // BUG CHECK: Ensure that the kernel has filled in the pre-packed weight to be cached if the weight was pre-packed
+                    ORT_ENFORCE(weights_to_be_filled_in.is_filled_, "The kernel corresponding to the node ", node.Name(),
+                                " doesn't have an implementation that can cache computed pre-packed weights");
+
+                    auto prepacked_weight_hash = weights_to_be_filled_in.GetHash();
+
+                    const auto& op_type = node.OpType();
+
+                    // Sanity check
+                    // TODO: Check if some version of the ONNX IR allows op_type to be empty
+                    ORT_ENFORCE(!op_type.empty(), "The op type of a node cannot be empty");
+
+                    std::ostringstream ss_1;
```

Comment:
Can we use the GenerateKey() method here? This way you can achieve hashing of both the pre-packed and the shared init (marked as TODO).

Reply:
Fixed based on offline sync